explorer/ast/value.cpp

```diff
@@ -376,7 +376,8 @@ static auto SetFieldImpl(
                                     path_end, field_value, source_loc);
           inits.ok()) {
         return arena->New<NominalClassValue>(
-            &object.type(), *inits, object.base(), object.class_value_ptr());
+            &object.type(), *inits, object.base(),
```

Comment:
To maintain the invariant, do we also need to recreate all the base classes with the new pointer?

Reply:
Maybe? Do you have a suggestion? I don't really know what is going on in this piece of code.

Reply:
The overall task here is to build a new value that is like `value` but has the element described by the `path` replaced by `field_value`. In this case, we're changing a field of a derived class but not changing the  base class, but we still need to update the base class (and its base class and so on) to hold the new pointer to the new derivedmost object.

github doesn't let me provide a suggested edit here, but I think we want something like this:

```
auto* class_value_ptr = arena->New<const NominalClassValue*>();
std::vector<const NominalClassValue*> base_path;
for (auto base = object.base(); base; base = (*base)->base()) {
  base_path.push_back(*base);
}
std::optional<Nonnull<const NominalClassValue*>> base;
for (auto* base_path_elem : llvm::reversed(base_path)) {
  base = arena->New<NominalClassValue>(
      base_path_elem->type(), base_path_elem->inits(), base, class_value_ptr);
}
return arena->New<NominalClassValue>(
    &object.type(), *inits, base, class_value_ptr);
```

(Longer term I'd love to get rid of `class_value_ptr` entirely, but I think we need to finish the expression category work first.)

Reply:
Great! With this fix, and the one other change from offline, the tests pass with the new assertion added.